src/hotspot/share/utilities/resourceHash.hpp

```diff
@@ -210,6 +211,32 @@ class ResourceHashtableBase : public STORAGE {
       ++bucket;
     }
   }
+
+  // ITER contains bool do_entry(K const&, V const&), which will be
+  // called for each entry in the table.  If do_entry() returns true,
+  // the entry is deleted.
+  template<class ITER>
+  void unlink(ITER* iter) const {
```

Comment:
`unlink()` shouldn't be `const` as it modifies the table. However, you probably needed to declare it as such because it calls `bucket_at()`, which is `const`. And the reason that  `bucket_at()` is `const` is because it's called by `iterate()`, which is `const`.

The use of `const` and `const_cast` in this class is getting a bit out of whack. I have created [JDK-8271525](https://bugs.openjdk.java.net/browse/JDK-8271525) with a preliminary fix here: https://github.com/openjdk/jdk/pull/4942

Maybe we should do that before this PR? it will also make the `const_cast` unnecessary for `decrement_entries()`.

Reply:
I agree with @tstuefe in your draft PR.  I don't think get should be non-const.  I haven't been able to parse through the monstrosity cast of lookup_node to know why and whether it should be changed.
I made unlink non-const and made the changes that he suggested there and that fixed the cast that I had to decrement the entries.